### ISSUE TITLE
Crash (use-after-free) when clicking "Undo" on GUI editor

### ISSUE THREAD

**Issue by mpsrig** [ffAudio/foleys_gui_magic#46]
When clicking undo on the GUI editor, if there is nothing to undo, the app crashes. I turned on ASAN (Address Sanitizer) and identified this as a use-after-free. Here are the ASAN logs:

```
=================================================================
==2574==ERROR: AddressSanitizer: heap-use-after-free on address 0x60800010f2a8 at pc 0x000116e14c61 bp 0x7ffee2b7dd00 sp 0x7ffee2b7dcf8
WRITE of size 4 at 0x60800010f2a8 thread T0
    #0 0x116e14c60 in juce::Atomic<int>::operator++() atomic:1020
    #1 0x116e7ab9f in juce::ReferenceCountedObject::incReferenceCount() juce_ReferenceCountedObject.h:76
    #2 0x1188c5917 in juce::ReferenceCountedObjectPtr<juce::ValueTree::SharedObject>::ReferenceCountedObjectPtr(juce::ValueTree::SharedObject&) juce_ReferenceCountedObject.h:274
    #3 0x11885d962 in juce::ReferenceCountedObjectPtr<juce::ValueTree::SharedObject>::ReferenceCountedObjectPtr(juce::ValueTree::SharedObject&) juce_ReferenceCountedObject.h:273
    #4 0x11885d877 in juce::ValueTree::ValueTree(juce::ValueTree::SharedObject&) juce_ValueTree.cpp:602
    #5 0x11885da02 in juce::ValueTree::ValueTree(juce::ValueTree::SharedObject&) juce_ValueTree.cpp:602
    #6 0x118868b45 in juce::ValueTree::SharedObject::getOrCreateChildWithName(juce::Identifier const&, juce::UndoManager*) juce_ValueTree.cpp:222
    #7 0x1188685d1 in juce::ValueTree::getOrCreateChildWithName(juce::Identifier const&, juce::UndoManager*) juce_ValueTree.cpp:895
    #8 0x116ffd7bd in foleys::MagicGUIBuilder::createDefaultGUITree(bool) foleys_MagicGUIBuilder.cpp:502
    #9 0x116fe19da in foleys::MagicGUIBuilder::updateComponents() foleys_MagicGUIBuilder.cpp:161
    #10 0x116fe0b3a in foleys::MagicGUIBuilder::updateAll() foleys_MagicGUIBuilder.cpp:65
    #11 0x117063b8b in foleys::MagicGUIBuilder::valueTreeChildRemoved(juce::ValueTree&, juce::ValueTree&, int) foleys_MagicGUIBuilder.cpp:564
    #12 0x1188b42b2 in juce::ValueTree::SharedObject::sendChildRemovedMessage(juce::ValueTree, int)::'lambda'(juce::ValueTree::Listener&)::operator()(juce::ValueTree::Listener&) const juce_ValueTree.cpp:113
    #13 0x1188b3ff5 in void juce::ListenerList<juce::ValueTree::Listener, juce::Array<juce::ValueTree::Listener*, juce::DummyCriticalSection, 0> >::callExcluding<juce::ValueTree::SharedObject::sendChildRemovedMessage(juce::ValueTree, int)::'lambda'(juce::ValueTree::Listener&)&>(juce::ValueTree::Listener*, juce::ValueTree::SharedObject::sendChildRemovedMessage(juce::ValueTree, int)::'lambda'(juce::ValueTree::Listener&)&&&) juce_ListenerList.h:140
    #14 0x1188b38d9 in void juce::ValueTree::SharedObject::callListeners<juce::ValueTree::SharedObject::sendChildRemovedMessage(juce::ValueTree, int)::'lambda'(juce::ValueTree::Listener&)>(juce::ValueTree::Listener*, juce::ValueTree::SharedObject::sendChildRemovedMessage(juce::ValueTree, int)::'lambda'(juce::ValueTree::Listener&)) const juce_ValueTree.cpp:75
    #15 0x1188b3444 in void juce::ValueTree::SharedObject::callListenersForAllParents<juce::ValueTree::SharedObject::sendChildRemovedMessage(juce::ValueTree, int)::'lambda'(juce::ValueTree::Listener&)>(juce::ValueTree::Listener*, juce::ValueTree::SharedObject::sendChildRemovedMessage(juce::ValueTree, int)::'lambda'(juce::ValueTree::Listener&)) const juce_ValueTree.cpp:95
    #16 0x1188b3094 in juce::ValueTree::SharedObject::sendChildRemovedMessage(juce::ValueTree, int) juce_ValueTree.cpp:113
    #17 0x118869cd2 in juce::ValueTree::SharedObject::removeChild(int, juce::UndoManager*) juce_ValueTree.cpp:297
    #18 0x1188af182 in juce::ValueTree::SharedObject::AddOrRemoveChildAction::undo() juce_ValueTree.cpp:508
    #19 0x11888352f in juce::UndoManager::ActionSet::undo() const juce_UndoManager.cpp:47
    #20 0x1188830ae in juce::UndoManager::undo() juce_UndoManager.cpp:252
    #21 0x11750ecb8 in foleys::ToolBox::ToolBox(juce::Component*, foleys::MagicGUIBuilder&)::$_43::operator()() const foleys_ToolBox.cpp:89
    #22 0x11750ebb5 in _ZNSt3__128__invoke_void_return_wrapperIvE6__callIJRZN6foleys7ToolBoxC1EPN4juce9ComponentERNS3_15MagicGUIBuilderEE4$_43EEEvDpOT_ type_traits:4428
    #23 0x11750e587 in std::__1::__function::__func<foleys::ToolBox::ToolBox(juce::Component*, foleys::MagicGUIBuilder&)::$_43, std::__1::allocator<foleys::ToolBox::ToolBox(juce::Component*, foleys::MagicGUIBuilder&)::$_43>, void ()>::operator()() functional:1562
    #24 0x1185ff965 in std::__1::function<void ()>::operator()() const functional:1913
    #25 0x1193c5e84 in juce::Button::sendClickMessage(juce::ModifierKeys const&) juce_Button.cpp:408
    #26 0x1193c7d2e in juce::Button::internalClickCallback(juce::ModifierKeys const&) juce_Button.cpp:351
    #27 0x1193ca13d in juce::Button::mouseUp(juce::MouseEvent const&) juce_Button.cpp:471
    #28 0x119345f69 in juce::Component::internalMouseUp(juce::MouseInputSource, juce::Point<float>, juce::Time, juce::ModifierKeys, float, float, float, float, float) juce_Component.cpp:2434
    #29 0x1199a561e in juce::MouseInputSourceInternal::sendMouseUp(juce::Component&, juce::Point<float>, juce::Time, juce::ModifierKeys) juce_MouseInputSource.cpp:155
    #30 0x119999dea in juce::MouseInputSourceInternal::setButtons(juce::Point<float>, juce::Time, juce::ModifierKeys) juce_MouseInputSource.cpp:197
    #31 0x119364f6a in juce::MouseInputSourceInternal::handleEvent(juce::ComponentPeer&, juce::Point<float>, juce::Time, juce::ModifierKeys, float, float, juce::PenDetails) juce_MouseInputSource.cpp:334
    #32 0x119363503 in juce::MouseInputSource::handleEvent(juce::ComponentPeer&, juce::Point<float>, long long, juce::ModifierKeys, float, float, juce::PenDetails const&) juce_MouseInputSource.cpp:634
    #33 0x1198c09d4 in juce::ComponentPeer::handleMouseEvent(juce::MouseInputSource::InputSourceType, juce::Point<float>, juce::ModifierKeys, float, float, long long, juce::PenDetails, int) juce_ComponentPeer.cpp:88
    #34 0x119ca40a9 in juce::NSViewComponentPeer::sendMouseEvent(NSEvent*) juce_mac_NSViewComponentPeer.mm:725
    #35 0x119ca5301 in juce::NSViewComponentPeer::redirectMouseUp(NSEvent*) juce_mac_NSViewComponentPeer.mm:607
    #36 0x119c92fc5 in juce::JuceNSViewClass::asyncMouseUp(objc_object*, objc_selector*, NSEvent*) juce_mac_NSViewComponentPeer.mm:1675
    #37 0x7fff2eb480b4 in __NSThreadPerformPerform (Foundation:x86_64+0x680b4)
    #38 0x7fff2ca07010 in __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ (CoreFoundation:x86_64+0xa0010)
    #39 0x7fff2cabe8eb in __CFRunLoopDoSource0 (CoreFoundation:x86_64+0x1578eb)
    #40 0x7fff2c9e9fbf in __CFRunLoopDoSources0 (CoreFoundation:x86_64+0x82fbf)
    #41 0x7fff2c9e943c in __CFRunLoopRun (CoreFoundation:x86_64+0x8243c)
    #42 0x7fff2c9e8c96 in CFRunLoopRunSpecific (CoreFoundation:x86_64+0x81c96)
    #43 0x7fff2bcc8d95 in RunCurrentEventLoopInMode (HIToolbox:x86_64+0x2fd95)
    #44 0x7fff2bcc8a0e in ReceiveNextEventCommon (HIToolbox:x86_64+0x2fa0e)
    #45 0x7fff2bcc8883 in _BlockUntilNextEventMatchingListInModeWithFilter (HIToolbox:x86_64+0x2f883)
    #46 0x7fff29f78a72 in _DPSNextEvent (AppKit:x86_64+0x41a72)
    #47 0x7fff2a70ee33 in -[NSApplication(NSEvent) _nextEventMatchingEventMask:untilDate:inMode:dequeue:] (AppKit:x86_64+0x7d7e33)
    #48 0x7fff29f6d884 in -[NSApplication run] (AppKit:x86_64+0x36884)
    #49 0x7fff29f3ca71 in NSApplicationMain (AppKit:x86_64+0x5a71)
    #50 0x7fff549b9014 in start (libdyld.dylib:x86_64+0x1014)

0x60800010f2a8 is located 8 bytes inside of 88-byte region [0x60800010f2a0,0x60800010f2f8)
freed by thread T0 here:
    #0 0x10d2d8d32 in wrap__ZdlPv (libclang_rt.asan_osx_dynamic.dylib:x86_64+0x61d32)
    #1 0x1188a12a7 in juce::ValueTree::SharedObject::~SharedObject() juce_ValueTree.cpp:51
    #2 0x1188a4ae9 in juce::ContainerDeletePolicy<juce::ValueTree::SharedObject>::destroy(juce::ValueTree::SharedObject*) juce_ContainerDeletePolicy.h:54
    #3 0x1188c565e in juce::ReferenceCountedObjectPtr<juce::ValueTree::SharedObject>::decIfNotNull(juce::ValueTree::SharedObject*) juce_ReferenceCountedObject.h:448
    #4 0x1188c55aa in juce::ReferenceCountedObjectPtr<juce::ValueTree::SharedObject>::~ReferenceCountedObjectPtr() juce_ReferenceCountedObject.h:374
    #5 0x11885ba5e in juce::ReferenceCountedObjectPtr<juce::ValueTree::SharedObject>::~ReferenceCountedObjectPtr() juce_ReferenceCountedObject.h:373
    #6 0x1188af297 in juce::ValueTree::SharedObject::AddOrRemoveChildAction::~AddOrRemoveChildAction() juce_ValueTree.cpp:476
    #7 0x1188ae96e in juce::ValueTree::SharedObject::AddOrRemoveChildAction::~AddOrRemoveChildAction() juce_ValueTree.cpp:476
    #8 0x1188ae9ce in juce::ValueTree::SharedObject::AddOrRemoveChildAction::~AddOrRemoveChildAction() juce_ValueTree.cpp:476
    #9 0x11887c2c9 in juce::UndoManager::perform(juce::UndoableAction*) memory:2285
    #10 0x1188625e2 in juce::ValueTree::SharedObject::addChild(juce::ValueTree::SharedObject*, int, juce::UndoManager*) juce_ValueTree.cpp:277
    #11 0x118868abf in juce::ValueTree::SharedObject::getOrCreateChildWithName(juce::Identifier const&, juce::UndoManager*) juce_ValueTree.cpp:221
    #12 0x1188685d1 in juce::ValueTree::getOrCreateChildWithName(juce::Identifier const&, juce::UndoManager*) juce_ValueTree.cpp:895
    #13 0x116ffd7bd in foleys::MagicGUIBuilder::createDefaultGUITree(bool) foleys_MagicGUIBuilder.cpp:502
    #14 0x116fe19da in foleys::MagicGUIBuilder::updateComponents() foleys_MagicGUIBuilder.cpp:161
    #15 0x116fe0b3a in foleys::MagicGUIBuilder::updateAll() foleys_MagicGUIBuilder.cpp:65
    #16 0x117063b8b in foleys::MagicGUIBuilder::valueTreeChildRemoved(juce::ValueTree&, juce::ValueTree&, int) foleys_MagicGUIBuilder.cpp:564
    #17 0x1188b42b2 in juce::ValueTree::SharedObject::sendChildRemovedMessage(juce::ValueTree, int)::'lambda'(juce::ValueTree::Listener&)::operator()(juce::ValueTree::Listener&) const juce_ValueTree.cpp:113
    #18 0x1188b3ff5 in void juce::ListenerList<juce::ValueTree::Listener, juce::Array<juce::ValueTree::Listener*, juce::DummyCriticalSection, 0> >::callExcluding<juce::ValueTree::SharedObject::sendChildRemovedMessage(juce::ValueTree, int)::'lambda'(juce::ValueTree::Listener&)&>(juce::ValueTree::Listener*, juce::ValueTree::SharedObject::sendChildRemovedMessage(juce::ValueTree, int)::'lambda'(juce::ValueTree::Listener&)&&&) juce_ListenerList.h:140
    #19 0x1188b38d9 in void juce::ValueTree::SharedObject::callListeners<juce::ValueTree::SharedObject::sendChildRemovedMessage(juce::ValueTree, int)::'lambda'(juce::ValueTree::Listener&)>(juce::ValueTree::Listener*, juce::ValueTree::SharedObject::sendChildRemovedMessage(juce::ValueTree, int)::'lambda'(juce::ValueTree::Listener&)) const juce_ValueTree.cpp:75
    #20 0x1188b3444 in void juce::ValueTree::SharedObject::callListenersForAllParents<juce::ValueTree::SharedObject::sendChildRemovedMessage(juce::ValueTree, int)::'lambda'(juce::ValueTree::Listener&)>(juce::ValueTree::Listener*, juce::ValueTree::SharedObject::sendChildRemovedMessage(juce::ValueTree, int)::'lambda'(juce::ValueTree::Listener&)) const juce_ValueTree.cpp:95
    #21 0x1188b3094 in juce::ValueTree::SharedObject::sendChildRemovedMessage(juce::ValueTree, int) juce_ValueTree.cpp:113
    #22 0x118869cd2 in juce::ValueTree::SharedObject::removeChild(int, juce::UndoManager*) juce_ValueTree.cpp:297
    #23 0x1188af182 in juce::ValueTree::SharedObject::AddOrRemoveChildAction::undo() juce_ValueTree.cpp:508
    #24 0x11888352f in juce::UndoManager::ActionSet::undo() const juce_UndoManager.cpp:47
    #25 0x1188830ae in juce::UndoManager::undo() juce_UndoManager.cpp:252
    #26 0x11750ecb8 in foleys::ToolBox::ToolBox(juce::Component*, foleys::MagicGUIBuilder&)::$_43::operator()() const foleys_ToolBox.cpp:89
    #27 0x11750ebb5 in _ZNSt3__128__invoke_void_return_wrapperIvE6__callIJRZN6foleys7ToolBoxC1EPN4juce9ComponentERNS3_15MagicGUIBuilderEE4$_43EEEvDpOT_ type_traits:4428
    #28 0x11750e587 in std::__1::__function::__func<foleys::ToolBox::ToolBox(juce::Component*, foleys::MagicGUIBuilder&)::$_43, std::__1::allocator<foleys::ToolBox::ToolBox(juce::Component*, foleys::MagicGUIBuilder&)::$_43>, void ()>::operator()() functional:1562
    #29 0x1185ff965 in std::__1::function<void ()>::operator()() const functional:1913

previously allocated by thread T0 here:
    #0 0x10d2d8752 in wrap__Znwm (libclang_rt.asan_osx_dynamic.dylib:x86_64+0x61752)
    #1 0x118868a07 in juce::ValueTree::SharedObject::getOrCreateChildWithName(juce::Identifier const&, juce::UndoManager*) juce_ValueTree.cpp:220
    #2 0x1188685d1 in juce::ValueTree::getOrCreateChildWithName(juce::Identifier const&, juce::UndoManager*) juce_ValueTree.cpp:895
    #3 0x116ffd7bd in foleys::MagicGUIBuilder::createDefaultGUITree(bool) foleys_MagicGUIBuilder.cpp:502
    #4 0x116fe19da in foleys::MagicGUIBuilder::updateComponents() foleys_MagicGUIBuilder.cpp:161
    #5 0x116fe0b3a in foleys::MagicGUIBuilder::updateAll() foleys_MagicGUIBuilder.cpp:65
    #6 0x117063b8b in foleys::MagicGUIBuilder::valueTreeChildRemoved(juce::ValueTree&, juce::ValueTree&, int) foleys_MagicGUIBuilder.cpp:564
    #7 0x1188b42b2 in juce::ValueTree::SharedObject::sendChildRemovedMessage(juce::ValueTree, int)::'lambda'(juce::ValueTree::Listener&)::operator()(juce::ValueTree::Listener&) const juce_ValueTree.cpp:113
    #8 0x1188b3ff5 in void juce::ListenerList<juce::ValueTree::Listener, juce::Array<juce::ValueTree::Listener*, juce::DummyCriticalSection, 0> >::callExcluding<juce::ValueTree::SharedObject::sendChildRemovedMessage(juce::ValueTree, int)::'lambda'(juce::ValueTree::Listener&)&>(juce::ValueTree::Listener*, juce::ValueTree::SharedObject::sendChildRemovedMessage(juce::ValueTree, int)::'lambda'(juce::ValueTree::Listener&)&&&) juce_ListenerList.h:140
    #9 0x1188b38d9 in void juce::ValueTree::SharedObject::callListeners<juce::ValueTree::SharedObject::sendChildRemovedMessage(juce::ValueTree, int)::'lambda'(juce::ValueTree::Listener&)>(juce::ValueTree::Listener*, juce::ValueTree::SharedObject::sendChildRemovedMessage(juce::ValueTree, int)::'lambda'(juce::ValueTree::Listener&)) const juce_ValueTree.cpp:75
    #10 0x1188b3444 in void juce::ValueTree::SharedObject::callListenersForAllParents<juce::ValueTree::SharedObject::sendChildRemovedMessage(juce::ValueTree, int)::'lambda'(juce::ValueTree::Listener&)>(juce::ValueTree::Listener*, juce::ValueTree::SharedObject::sendChildRemovedMessage(juce::ValueTree, int)::'lambda'(juce::ValueTree::Listener&)) const juce_ValueTree.cpp:95
    #11 0x1188b3094 in juce::ValueTree::SharedObject::sendChildRemovedMessage(juce::ValueTree, int) juce_ValueTree.cpp:113
    #12 0x118869cd2 in juce::ValueTree::SharedObject::removeChild(int, juce::UndoManager*) juce_ValueTree.cpp:297
    #13 0x1188af182 in juce::ValueTree::SharedObject::AddOrRemoveChildAction::undo() juce_ValueTree.cpp:508
    #14 0x11888352f in juce::UndoManager::ActionSet::undo() const juce_UndoManager.cpp:47
    #15 0x1188830ae in juce::UndoManager::undo() juce_UndoManager.cpp:252
    #16 0x11750ecb8 in foleys::ToolBox::ToolBox(juce::Component*, foleys::MagicGUIBuilder&)::$_43::operator()() const foleys_ToolBox.cpp:89
    #17 0x11750ebb5 in _ZNSt3__128__invoke_void_return_wrapperIvE6__callIJRZN6foleys7ToolBoxC1EPN4juce9ComponentERNS3_15MagicGUIBuilderEE4$_43EEEvDpOT_ type_traits:4428
    #18 0x11750e587 in std::__1::__function::__func<foleys::ToolBox::ToolBox(juce::Component*, foleys::MagicGUIBuilder&)::$_43, std::__1::allocator<foleys::ToolBox::ToolBox(juce::Component*, foleys::MagicGUIBuilder&)::$_43>, void ()>::operator()() functional:1562
    #19 0x1185ff965 in std::__1::function<void ()>::operator()() const functional:1913
    #20 0x1193c5e84 in juce::Button::sendClickMessage(juce::ModifierKeys const&) juce_Button.cpp:408
    #21 0x1193c7d2e in juce::Button::internalClickCallback(juce::ModifierKeys const&) juce_Button.cpp:351
    #22 0x1193ca13d in juce::Button::mouseUp(juce::MouseEvent const&) juce_Button.cpp:471
    #23 0x119345f69 in juce::Component::internalMouseUp(juce::MouseInputSource, juce::Point<float>, juce::Time, juce::ModifierKeys, float, float, float, float, float) juce_Component.cpp:2434
    #24 0x1199a561e in juce::MouseInputSourceInternal::sendMouseUp(juce::Component&, juce::Point<float>, juce::Time, juce::ModifierKeys) juce_MouseInputSource.cpp:155
    #25 0x119999dea in juce::MouseInputSourceInternal::setButtons(juce::Point<float>, juce::Time, juce::ModifierKeys) juce_MouseInputSource.cpp:197
    #26 0x119364f6a in juce::MouseInputSourceInternal::handleEvent(juce::ComponentPeer&, juce::Point<float>, juce::Time, juce::ModifierKeys, float, float, juce::PenDetails) juce_MouseInputSource.cpp:334
    #27 0x119363503 in juce::MouseInputSource::handleEvent(juce::ComponentPeer&, juce::Point<float>, long long, juce::ModifierKeys, float, float, juce::PenDetails const&) juce_MouseInputSource.cpp:634
    #28 0x1198c09d4 in juce::ComponentPeer::handleMouseEvent(juce::MouseInputSource::InputSourceType, juce::Point<float>, juce::ModifierKeys, float, float, long long, juce::PenDetails, int) juce_ComponentPeer.cpp:88
    #29 0x119ca40a9 in juce::NSViewComponentPeer::sendMouseEvent(NSEvent*) juce_mac_NSViewComponentPeer.mm:725

SUMMARY: AddressSanitizer: heap-use-after-free atomic:1020 in juce::Atomic<int>::operator++()
```

**Comment by ffAudio** [comment]
Thanks for the report. I have hit this sporadically, so I am aware.
Currently I am in the middle of a major refactor that should simplify the update mechanisms, so I would look into that after the refactor, since it is avoidable (even though annoying, since sometimes you don't know when to stop undoing...).
I have (though little) hope, that the refactored version doesn't suffer from that.

**Comment by ffAudio** [comment]
As far as I can see it is fixed now in version 1.2.0 of 18/06/2020
Please let me know if the error persists, I can no longer reproduce it.

**Comment by ffAudio** [comment]
Closing since it seems to be fixed. If it still happens, please reopen.
